src/requirements/requirement-graph-builder.ts

```diff
@@ -0,0 +1,122 @@
+  /** See RequirementFulfillmentGraph for its usage. */
+  readonly getCourseUniqueID: (c: Course) => string;
+  /**
+   * Aaively give a list of courses that can satisfy a requirement. Most of the time this function
+   * should just return the pre-computed course list. For requirements have multiple fulfillment
+   * strategies, it will return the union of all pre-computed course list.
+   */
+  readonly getAllCoursesThatCanPotentiallySatisfyRequirement: (
+    requirement: Requirement
+  ) => readonly Course[];
+  /**
+   * For a requirement with multiple fulfillment strategies, this function should return the list of
+   * courses bind to one fulfillment strategies.
+   */
+  readonly getCorrespondingRequirementAndAllRelevantCoursesUnderFulfillmentStrategy: (
+    choice: UserChoiceOnFulfillmentStrategy
+  ) => {
+    readonly correspondingRequirement: Requirement;
+    readonly coursesOfChosenFulfillmentStrategy: Course[];
+  };
+};
+
+const buildRequirementFulfillmentGraph = <Requirement, Course, UserChoiceOnFulfillmentStrategy>({
+  requirements,
+  userCourses,
+  userChoiceOnFulfillmentStrategy,
+  userChoiceOnDoubleCountingElimiation,
+  getRequirementUniqueID,
+  getCourseUniqueID,
+  getAllCoursesThatCanPotentiallySatisfyRequirement,
+  getCorrespondingRequirementAndAllRelevantCoursesUnderFulfillmentStrategy
+}: BuildRequirementFulfillmentGraphParameters<
+  Requirement,
+  Course,
+  UserChoiceOnFulfillmentStrategy
+>): RequirementFulfillmentGraph<Requirement, Course> => {
+  const graph = new RequirementFulfillmentGraph(getRequirementUniqueID, getCourseUniqueID);
+  const userCourseSet = new HashMap<Course, null>(getCourseUniqueID);
+  userCourses.forEach(course => userCourseSet.set(course, null));
+
+  // Phase 1:
+  //   Building a rough graph by naively connecting requirements and courses based on
+  //   `getAllCoursesThatCanPotentiallySatisfyRequirement`.
+  requirements.forEach(requirement => {
+    getAllCoursesThatCanPotentiallySatisfyRequirement(requirement).forEach(course => {
+      if (userCourseSet.has(course)) graph.addEdge(requirement, course);
+    });
+  });
+
+  // Phase 2-1: Respect user's choices on fulfillment strategies.
+  userChoiceOnFulfillmentStrategy.forEach(choice => {
+    const {
+      correspondingRequirement,
+      coursesOfChosenFulfillmentStrategy
+    } = getCorrespondingRequirementAndAllRelevantCoursesUnderFulfillmentStrategy(choice);
+    const coursesToKeepSet = new HashMap<Course, null>(getCourseUniqueID);
+    coursesOfChosenFulfillmentStrategy.forEach(course => coursesToKeepSet.set(course, null));
+
+    graph.getConnectedCoursesFromRequirement(correspondingRequirement).forEach(connectedCourse => {
+      if (!coursesToKeepSet.has(connectedCourse)) {
+        graph.removeEdge(correspondingRequirement, connectedCourse);
+      }
+    });
+  });
+
+  // Phase 2-2:
+  //   Auto-select fulfillment for requirements with multiple fulfillment strategies
```

Comment:
Based on the current WIP design, I think users will have to select a fulfillment before adding a course on the add course modal. This modal does have default fulfillments pre-selected in the design, so would we just want default generating code there instead of auto-selecting code here?  Let me know if that makes sense or if I'm off about anything.

Reply:
if we want  default generating code then it will not be part of this graph builder phase. we can do it elsewhere.